adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala

```diff
@@ -44,86 +46,120 @@ class FeatureFile(parser: FeatureParser) extends Serializable {
 }
 
 object GTFParser {
+  private val PATTERN = "\\s*([^\\s]+)\\s\"([^\"]+)\"".r
 
-  private val attr_regex = "\\s*([^\\s]+)\\s\"([^\"]+)\"".r
-
-  /**
-   * Parses a string of format
-   *   token; token; token ...
-   *
-   * where each 'token' is of the form
-   *   key "value"
-   *
-   * and turns it into a Map
-   *
-   * @param attributeField The original string of tokens
-   * @return The Map of attributes
-   */
-  def parseAttrs(attributeField: String): Map[String, String] =
+  def parseAttributes(attributeField: String): Seq[(String, String)] =
     attributeField.split(";").flatMap {
       case token: String =>
-        attr_regex.findFirstMatchIn(token).map(m => (m.group(1), m.group(2)))
-    }.toMap
+        PATTERN.findFirstMatchIn(token).map(m => (m.group(1), m.group(2)))
+    }
 }
 
 /**
- * GTF is a line-based GFF variant.
+ * Parser for GTF/GFF2 format.
  *
- * Details of the GTF/GFF format here:
+ * Specification:
  * http://www.ensembl.org/info/website/upload/gff.html
  */
 class GTFParser extends FeatureParser {
 
   override def parse(line: String): Seq[Feature] = {
-    // Just skip the '#' prefixed lines, these are comments in the
-    // GTF file format.
+    // skip header and comment lines
     if (line.startsWith("#")) {
       return Seq()
     }
 
     val fields = line.split("\t")
-
-    val (seqname, source, feature, start, end, score, strand, frame, attribute) =
+    // skip empty or invalid lines
+    if (fields.length < 9) {
+      return Seq()
+    }
+    val (seqname, source, featureType, start, end, score, strand, frame, attributes) =
       (fields(0), fields(1), fields(2), fields(3), fields(4), fields(5), fields(6), fields(7), fields(8))
 
-    lazy val attrs = GTFParser.parseAttrs(attribute)
-
-    val contig = Contig.newBuilder().setContigName(seqname).build()
     val f = Feature.newBuilder()
-      .setContig(contig)
-      .setStart(start.toLong - 1) // GTF/GFF ranges are 1-based
-      .setEnd(end.toLong) // GTF/GFF ranges are closed
-      .setFeatureType(feature)
       .setSource(source)
+      .setFeatureType(featureType)
+      .setContigName(seqname)
+      .setStart(start.toLong - 1L) // GTF/GFF2 coordinate system is 1-based
+      .setEnd(end.toLong) // GTF/GFF2 ranges are closed
+
+    // set score if specified
+    if (score != ".") {
+      f.setScore(source.toDouble)
+    }
+    // set frame if specified
+    if (frame != ".") {
+      f.setFrame(frame.toInt)
+    }
+    // set strand if specified
+    Features.toStrand(strand).foreach(f.setStrand(_))
 
-    val _strand = strand match {
-      case "+" => Strand.Forward
-      case "-" => Strand.Reverse
-      case _   => Strand.Independent
+    // assign values for various feature fields from attributes
+    Features.assignAttributes(GTFParser.parseAttributes(attributes), f)
+
+    Seq(f.build())
+  }
+}
+
+object GFF3Parser {
+  def parseAttributes(attributes: String): Seq[(String, String)] =
```

Comment:
Add docs? This is a weird line. I don't love the pattern match on the array, but that's just me. I would probably do something like:

```
attributes.split(";")
  .map(s => {
    val eqIdx = s.indexOf("=")
    (s.take(eqIdx), s.drop(eqIdx + 1))
  }).toSeq
```